afl-fuzz.c

```diff
@@ -2499,6 +2500,30 @@ char ReadCommandFromPipe(u32 timeout)
 	return result;
 }
 
+int ReadDWORDFromPipe(u32 timeout, DWORD *result)
+{
+	DWORD num_read;
+	//char result = 0;
+	if (!is_child_running())
+	{
+		return 0;
+	}
+
+	if (ReadFile(pipe_handle, result, 4, &num_read, &pipe_overlapped) || GetLastError() == ERROR_IO_PENDING)
```

Comment:
I don't think ReadFile here guarantees it will read all 4 bytes in one call (it's easier with ReadCommandFromPipe when only 1 byte needs to be read). I suggest something like

```
DWORD read_so_far = 0;
while(num_read_total < sizeof(DWORD)) {
	if (ReadFile(pipe_handle, (char*)result + read_so_far, sizeof(DWORD) - read_so_far, &num_read, &pipe_overlapped) || GetLastError() == ERROR_IO_PENDING)
	{
		//ACTF("ReadFile success or GLE IO_PENDING", result);
		if (WaitForSingleObject(pipe_overlapped.hEvent, timeout) != WAIT_OBJECT_0) {
			// took longer than specified timeout or other error - cancel read
			CancelIo(pipe_handle);
			WaitForSingleObject(pipe_overlapped.hEvent, INFINITE); //wait for cancelation to finish properly.
			*result = 0;
                        break;
		}
	}
}
```

Reply:
```suggestion
	while (read_so_far < sizeof(DWORD)) {
		PeekNamedPipe(pipe_handle, &result + read_so_far, sizeof(DWORD) - read_so_far, &num_read, &total_available, &byte_left);
		read_so_far += num_read;

		}
```
insteal of readfile, how about PeekNamedPipe function? i found some issue with ReadFile, with above changes you suggested it basically tries to read the pipe and sometimes returns 0 so exception code is not available. but with PeekNamedPipe its working fine. i am not expert in this so just asking before i sent out pull request.

Reply:
This wouldn't work because PeekNamedPipe won't remove data from the pipe, so if there is only 1 byte available at the time of the read, the above code will just copy that 1 byte in the entire output. You could PeekNamedPipe until there are at least 4 bytes available but this could result in (a) wasting resources and (b) potentially infinite loop.

So I'd rather find a way with ReadFile, in the case you mentioned when it returns 0, what does the GetLastError() return?

Reply:
there are two error codes i am getting:
1. 109 which  means ERROR_BROKEN_PIPE
2. 997 which means IO pending.

what i think is happening is that when winafl.c write command 'C' to the pipe, it then writes 4 bytes exception code to the pipe.
in the afl-fuzz.c it first reads 'C' command and after that when it tries to read the exception code, it gets one of the above error. and thus crash code is not available in some cases. any ideas you can suggest for this?

when i changed the buffer to 5 bytes, it worked fine because exception codes were transferred with the commands so no issues there.

Reply:
Interesting, I guess that is because by the time of the second read the client has already been killed and disconnected from the pipe. Maybe I'm wrong, but I would expect it should still be possible to read from the pipe even after the client disconnects - does any data get read at all in this case (num_read > 0)? And in your experiments with PeekNamedPipe did the data get read correctly?

Reply:
for the first time num_read ==0 in some case, and then its reading 4 bytes properly. this 0 is occurring intermittently.
i did run the build with PeekNamedPipe for few hours with 1 master and 7 slaves and it worked perfectly. i verified the crashes with another scripts and they were having correct crash codes(well all had the same crash code actually). although documentation says that data wont be removed from Pipe in case of PeekNamedPipe so i am worried if it breaks in some conditions?
I think we can do following:
1. change the command to 5 bytes long. or
2. change it to PeekNamedPipe(but with the cocern that will it work perfectly or not) or
3. use ReadFile which in some cases wont give EXCEPTION_CODE_NOT_AVAILABLE.

will you be able to spare few mins and check or debug readfile approach? is there anything we can do?

Reply:
Random thought: I wonder if the reason for ERROR_BROKEN_PIPE is that we aren't closing the pipe properly after sending the 'C' command and the crash code. Could you try adding `CloseHandle(pipe);` after sending the crash code?

Reply:
finally found kind of a heck. adding Sleep(0) works with the changes as you requested before:
following is the code, let me know if everything is fine and i will send pull request.

`Sleep(0);
	while (read_so_far < sizeof(DWORD)) {
		if (ReadFile(pipe_handle, &result + read_so_far, sizeof(DWORD) - read_so_far, &num_read, &pipe_overlapped) || GetLastError() == ERROR_IO_PENDING)
		{
			read_so_far += num_read;
			if (read_so_far >= sizeof(DWORD)) {
				break;
			}
			
			//ACTF("ReadFile success or GLE IO_PENDING");
			if (WaitForSingleObject(pipe_overlapped.hEvent, timeout) != WAIT_OBJECT_0) {
				// took longer than specified timeout or other error - cancel read
				CancelIo(pipe_handle);
				WaitForSingleObject(pipe_overlapped.hEvent, INFINITE); //wait for cancelation to finish properly.
				result = 0;
			}
		}
	}`

Reply:
Hm yeah that's a bit hacky (also no idea why it would work). Any luck with CloseHandle?

Reply:
tried closehandle too. no luck i think the read in afl-fuzz.c is too fast. adding sleep may cause some delay and so the read is successful. are you ok with this change? i think 1 ms should be fine?

Reply:
Give me some time to play with this some. I don't think calling ReadFile too fast should be the issue (after all, afl-fuzz calls ReadFile to get the status char from the client before that status is even known), but I don't know what the issue is. Let me take a look and report back.

Reply:
sure ivan.  i will wait for your confirmation.

Reply:
Ok, I think I see where the issue is. It seems that, with the overlapped pipe, num_read is always going to be zero, so the loop that I proposed doesn't make sense. However, in all my tests, all 4 bytes were in fact successful read. So, errm, I actually think that your initial code (calling ReadFile once) works best here. Sorry for the confusion! :-/ I'm going to mark this comment as resolved. Once you address the other comments, I'll merge the patch.